pkg/ccl/backupccl/alter_backup_planning.go

```diff
@@ -0,0 +1,265 @@
+			if len(backup) != 1 {
+				return errors.Errorf("ALTER BACKUP ... IN can only by used against a single collection path (per-locality)")
+			}
+
+			appendPaths := func(uris []string, tailDir string) error {
+				for i, uri := range uris {
+					parsed, err := url.Parse(uri)
+					if err != nil {
+						return err
+					}
+					parsed.Path = path.Join(parsed.Path, tailDir)
+					uris[i] = parsed.String()
+				}
+				return nil
+			}
+
+			if err = appendPaths(backup, subdir); err != nil {
+				return err
+			}
+		}
+
+		var newKms []string
+		newKms, err = newKmsFn()
+		if err != nil {
+			return err
+		}
+
+		var oldKms []string
+		oldKms, err = oldKmsFn()
+		if err != nil {
+			return err
+		}
+
+		return doAlterBackupPlan(ctx, alterBackupStmt, p, backup, newKms, oldKms)
+	}
+
+	return fn, nil, nil, false, nil
+}
+
+func doAlterBackupPlan(
+	ctx context.Context,
+	alterBackupStmt *tree.AlterBackup,
+	p sql.PlanHookState,
+	backup []string,
+	newKms []string,
+	oldKms []string,
+) error {
+	if len(backup) < 1 || len(backup[0]) < 1 {
+		return errors.New("invalid base backup specified")
+	}
+
+	baseStores, err := p.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx, backup[0], p.User())
+	if err != nil {
+		return errors.Wrapf(err, "failed to open backup storage location")
+	}
+	defer baseStores.Close()
+
+	opts, err := readEncryptionOptions(ctx, baseStores)
+	if err != nil {
+		return err
+	}
+
+	ioConf := baseStores.ExternalIOConf()
+
+	// Check that at least one of the old keys works on the backup.
+	// Use the first one that works to decrypt the ENCRYPTION-INFO file(s).
+	var defaultKMSInfo *jobspb.BackupEncryptionOptions_KMSInfo
+	oldKMSFound := false
+	for i := len(oldKms) - 1; i >= 0; i-- {
+		for _, encFile := range opts {
+			defaultKMSInfo, err = validateKMSURIsAgainstFullBackup([]string{oldKms[i]},
+				newEncryptedDataKeyMapFromProtoMap(encFile.EncryptedDataKeyByKMSMasterKeyID), &backupKMSEnv{
+					baseStores.Settings(),
+					&ioConf,
+				})
+
+			if err == nil {
+				oldKMSFound = true
+				break
+			}
+		}
+		if oldKMSFound {
+			break
+		}
+	}
+	if !oldKMSFound {
+		return fmt.Errorf("no key in old_kms matches the provided backup")
+	}
+
+	encryption := &jobspb.BackupEncryptionOptions{
+		Mode:    jobspb.EncryptionMode_KMS,
+		KMSInfo: defaultKMSInfo}
+
+	// Recover the encryption key using the old key, so we can encrypt it again with the new keys.
+	var plaintextDataKey []byte
+	plaintextDataKey, err = getEncryptionKey(ctx, encryption, baseStores.Settings(),
+		baseStores.ExternalIOConf())
+
+	if err != nil {
+		return err
+	}
+
+	kmsEnv := &backupKMSEnv{settings: p.ExecCfg().Settings, conf: &p.ExecCfg().ExternalIODirConfig}
+
+	encryptedDataKeyByKMSMasterKeyID := newEncryptedDataKeyMap()
+
+	// Add each new key user wants to add to a new data key map.
+	for _, kmsURI := range newKms {
+		masterKeyID, encryptedDataKey, err := getEncryptedDataKeyFromURI(ctx,
+			plaintextDataKey, kmsURI, kmsEnv)
+		if err != nil {
+			return err
```

Comment:
maybe wrap this error so that its more informative when it bubble up. Something like:

return errors.Wrap(err, "failed to encrypt data key when adding new KMS")